### PR TITLE
[mxfp8 training] refactor MXLinear config with separate options for dim0 and dim1 quantization kernel choice

### PR DESCRIPTION
**Pull request by danielvegamyhre**
Stacked PRs:
 * #3569
 * __->__#3568

--- --- ---

### Summary
- refactor MXLinearConfig with separate options for dim0 and dim1 quantization kernel choice
- refactor MXTensor with config for dim0 cast kernel
- this is important since torch.compile is not fast for dim0 + RCEIL rounding yet, and this is the default rounding mode we are migrating to. 
- Update validation so tests pass for RCEIL

### Tests
- `pytest test/prototype/mx_formats/test_mx_linear.py`

### Kernel microbenchmarks
- to_mx + torch.compile = ~3.7 TB/s
- triton = ~5.8 TB/s 
See: https://github.com/pytorch/ao/tree/main/torchao/prototype/mx_formats#quantization-kernel-microbenchmarks

### e2e 8 layer llama3 training benchmarks with torchtitan, fsdp=2, compile=true
- before (torch dim0 rceil quant):  78756.0 TPS
- after (triton dim0 rceil quant): 81144.5 TPS (3% boost)

### Benchmarks (e2e linear sigmoid fwd + bwd)

NO FUSION: 

Command: `python benchmarks/float8/float8_roofline.py --mx_recipe_name="mxfp8_cublas_rceil" --outfile "outfile.csv"`

Before:
```
   fwd_M  fwd_K  fwd_N  r_bf16_gemm_s  r_fp8_gemm_s  r_fp8_ovhd_s  ...  b_fp8_gemm_s  b_bf16_e2e_s  b_fp8_e2e_s  b_fp8_e2e_spdp  rb_bf16_gemm_ratio  rb_fp8_gemm_ratio
0   1024   1024   1024       6.00e-06      6.00e-06      1.20e-05  ...      1.77e-05      1.49e-05     4.87e-05            0.31                0.41               0.34
1   2048   2048   2048       2.94e-05      1.47e-05      1.20e-05  ...      3.52e-05      4.56e-05     9.29e-05            0.49                0.65               0.42
2   4096   4096   4096       2.35e-04      1.17e-04      4.10e-05  ...      1.37e-04      2.80e-04     3.12e-04            0.90                0.82               0.85
3   8192   8192   8192       1.88e-03      9.40e-04      1.64e-04  ...      8.34e-04      1.92e-03     1.56e-03            1.23                0.97               1.13
4  16384  16384  16384       1.50e-02      7.52e-03      6.56e-04  ...      9.34e-03      1.65e-02     1.30e-02            1.27                0.96               0.81
```

After:
```
fwd_M  fwd_K  fwd_N  r_bf16_gemm_s  r_fp8_gemm_s  r_fp8_ovhd_s  ...  b_fp8_gemm_s  b_bf16_e2e_s  b_fp8_e2e_s  b_fp8_e2e_spdp  rb_bf16_gemm_ratio  rb_fp8_gemm_ratio
0   1024   1024   1024       6.00e-06      6.00e-06      1.20e-05  ...      1.75e-05      1.49e-05     4.49e-05            0.33                0.41               0.34
1   2048   2048   2048       2.94e-05      1.47e-05      1.20e-05  ...      3.53e-05      4.54e-05     7.55e-05            0.60                0.65               0.42
2   4096   4096   4096       2.35e-04      1.17e-04      4.10e-05  ...      1.37e-04      2.79e-04     2.44e-04            1.14                0.82               0.86
3   8192   8192   8192       1.88e-03      9.40e-04      1.64e-04  ...      8.35e-04      1.91e-03     1.29e-03            1.48                0.97               1.13
4  16384  16384  16384       1.50e-02      7.52e-03      6.56e-04  ...      9.22e-03      1.72e-02     1.19e-02            1.45                0.98               0.82

```

WITH FUSION:

Before:
```
(torch) dev@gpu-dev-eef72896:~/ao$ python benchmarks/float8/float8_roofline.py --mx_recipe_name="mxfp8_cublas_rceil" --outfile "outfile.csv" --enable_fusion_modeling

   fwd_M  fwd_K  fwd_N  r_bf16_gemm_s  r_fp8_gemm_s  r_fp8_ovhd_s  ...  b_fp8_gemm_s  b_bf16_e2e_s  b_fp8_e2e_s  b_fp8_e2e_spdp  rb_bf16_gemm_ratio  rb_fp8_gemm_ratio
0   1024   1024   1024       6.00e-06      6.00e-06      1.20e-05  ...      1.75e-05      2.38e-05     5.54e-05            0.43                0.42               0.34
1   2048   2048   2048       2.94e-05      1.47e-05      1.20e-05  ...      3.47e-05      6.92e-05     1.13e-04            0.61                0.65               0.42
2   4096   4096   4096       2.35e-04      1.17e-04      3.19e-05  ...      1.37e-04      3.47e-04     3.66e-04            0.95                0.82               0.86
3   8192   8192   8192       1.88e-03      9.40e-04      1.28e-04  ...      8.41e-04      2.11e-03     1.78e-03            1.19                0.97               1.12
4  16384  16384  16384       1.50e-02      7.52e-03      5.11e-04  ...      9.05e-03      1.72e-02     1.29e-02            1.33                0.98               0.83

```

After:
```
(torch) dev@gpu-dev-eef72896:~/ao$ python benchmarks/float8/float8_roofline.py --mx_recipe_name="mxfp8_cublas_rceil" --outfile "outfile.csv" --enable_fusion_modeling

   fwd_M  fwd_K  fwd_N  r_bf16_gemm_s  r_fp8_gemm_s  r_fp8_ovhd_s  ...  b_fp8_gemm_s  b_bf16_e2e_s  b_fp8_e2e_s  b_fp8_e2e_spdp  rb_bf16_gemm_ratio  rb_fp8_gemm_ratio
0   1024   1024   1024       6.00e-06      6.00e-06      1.20e-05  ...      1.75e-05      2.38e-05     5.49e-05            0.43                0.41               0.34
1   2048   2048   2048       2.94e-05      1.47e-05      1.20e-05  ...      2.77e-05      6.70e-05     9.67e-05            0.69                0.81               0.53
2   4096   4096   4096       2.35e-04      1.17e-04      3.19e-05  ...      1.09e-04      3.46e-04     3.03e-04            1.14                1.03               1.07
3   8192   8192   8192       1.88e-03      9.40e-04      1.28e-04  ...      6.75e-04      2.11e-03     1.53e-03            1.38                1.21               1.39
4  16384  16384  16384       1.50e-02      7.52e-03      5.11e-04  ...      7.31e-03      1.74e-02     1.21e-02            1.44                1.24               1.03
```

### CODE REVIEW

**Comment by vkuzo** [comment]
what do benchmarks look like

**Comment by danielvegamyhre** [comment]
> what do benchmarks look like

Updated PR description with benchmarks, they are better

**Comment by vkuzo** [comment]
> Updated PR description with benchmarks, they are better

that looks good for no-fusion, do we have metrics on what this looks like with layer fusion?  The script has a flag to enable it, or you could use LLaMa 3 8B in torchtitan.

File: torchao/prototype/mx_formats/mx_linear.py [hunk 1]
```diff
@@ -61,19 +62,19 @@ def forward(
         input_orig_shape = input_hp.shape
         input_hp_r = input_hp.reshape(-1, input_orig_shape[-1])
 
-        input_mx_r_dim0 = MXTensor.to_mx(
+        input_mx_r_dim0 = _to_mxfp8_dim0(
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_linear.py]: IMO we should ensure the codebase stays consistent with how the various config flags are used.  Before this PR, `mxfp8_cast_kernel_choice` controls the dim1 cast.  I think good options are:
1. make `mxfp8_cast_kernel_choice` control dim1 and dim0 casts (update names/docs as needed)
2. or, make a separate `mxfp8_cast_kernel_choice` thing for dim0, update names as needed
3. or, just delete all the configs and make the logic use custom kernels

The current PR is a bit confusing given the rest of the feature because it leaves the dim1 config as is but defaults the dim0 logic to a custom kernel.

I don't really like Option 3 because we should be able to quickly test compile, so Option 1 or Option 2 would sound good to me.

File: torchao/prototype/mx_formats/mx_linear.py [hunk 2]
```diff
@@ -231,3 +232,42 @@ def extra_repr(self):
...
+    kernel_preference: KernelPreference,
+    scale_calculation_mode: ScaleCalculationMode,
+) -> MXTensor:
+    """
+    Quantize a tensor to MXFP8 along dimension 0.
+
+    Uses Triton kernel for RCEIL scaling mode for better performance, otherwise
+    falls back to MXTensor.to_mx().
+    """
+    if scale_calculation_mode == ScaleCalculationMode.RCEIL:
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_linear.py]: this isn't how the dim1 config works, let's make sure all of these flags operate in a consistent way.  You can change the dim1 logic if needed (everything is prototype).

File: torchao/prototype/mx_formats/mx_linear.py (hunk 1 above)
Comment (danielvegamyhre) [reply torchao/prototype/mx_formats/mx_linear.py]: Makes sense, I opted to do option 2, since dim0 and dim1 have different supported kernels atm (torch/triton for dim0, torch/triton/cuda for dim1). Refactored and updated tests acccordingly.

This is ready for another pass when you have time.

File: torchao/prototype/mx_formats/mx_linear.py [hunk 3]
```diff
@@ -231,3 +240,43 @@ def extra_repr(self):
 @register_quantize_module_handler(MXLinearConfig)
 def _mx_linear_transform(module: torch.nn.Module, config: MXLinearConfig):
     return MXLinear.from_float(module, config=config)
+
+
+def _to_mxfp8_dim0(
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_linear.py]: this seems to logically belong in `mx_tensor.py`, so it can be reused by other callsites such as inference

**Comment by vkuzo** [comment]
optionally, would be good to split the rename of the existing name in a separate PR from adding the new thing

File: torchao/prototype/mx_formats/mx_linear.py (hunk 3 above)
Comment (danielvegamyhre) [reply torchao/prototype/mx_formats/mx_linear.py]: Makes sense, moved it to mx_tensor.py

**Comment by danielvegamyhre** [comment]
> > Updated PR description with benchmarks, they are better
> 
> that looks good for no-fusion, do we have metrics on what this looks like with layer fusion? The script has a flag to enable it, or you could use LLaMa 3 8B in torchtitan.

Updated PR description with before/after benchmarks with fusion enabled, also look good

File: torchao/prototype/mx_formats/mx_tensor.py [hunk 4]
```diff
@@ -921,3 +925,43 @@ def mx_wait_tensor(func, types, args, kwargs):
...
+    scale_calculation_mode: ScaleCalculationMode,
+    mxfp8_dim0_cast_kernel_choice: MXFP8Dim0CastKernelChoice,
+) -> MXTensor:
+    """
+    Quantize a tensor to MXFP8 along dimension 0.
+
+    Uses Triton kernel or falls back to MXTensor.to_mx() based on the
+    mxfp8_dim0_cast_kernel_choice parameter.
+    """
+    if mxfp8_dim0_cast_kernel_choice == MXFP8Dim0CastKernelChoice.TRITON:
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_tensor.py]: how about putting this into `MXTensor.to_mx` instead? there are already too many different ways to create an MX tensor, let's not add one more without a good reason

Comment (danielvegamyhre) [reply torchao/prototype/mx_formats/mx_tensor.py]: yeah makes sense, did this and updated call sites accordingly. verified tests still pass.

File: torchao/prototype/mx_formats/mx_tensor.py [hunk 5]
```diff
@@ -540,10 +544,35 @@ def to_mx(
...
-            data_hp, elem_dtype, block_size, scaling_mode, is_swizzled_scales
+        assert mxfp8_dim0_cast_kernel_choice in (
+            MXFP8Dim0CastKernelChoice.TRITON,
+            MXFP8Dim0CastKernelChoice.TORCH,
+        ), (
+            f"unsupported kernel choice for mxfp8_dim0_cast_kernel_choice: {mxfp8_dim0_cast_kernel_choice}"
+        )
+
+        triton_kernel_supported = (
+            mxfp8_dim0_cast_kernel_choice == MXFP8Dim0CastKernelChoice.TRITON
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_tensor.py]: this is redundant, remove

File: torchao/prototype/mx_formats/mx_tensor.py [hunk 6]
```diff
@@ -540,10 +544,35 @@ def to_mx(
...
+            and data_hp.dtype == torch.bfloat16
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_tensor.py]: isn't this checked in the kernel? if yes, remove from here

File: torchao/prototype/mx_formats/mx_tensor.py [hunk 7]
```diff
@@ -540,10 +544,35 @@ def to_mx(
...
+            and elem_dtype == torch.float8_e4m3fn
+            and scaling_mode in (ScaleCalculationMode.FLOOR, ScaleCalculationMode.RCEIL)
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_tensor.py]: check in the kernel instead of here

**Review by vkuzo — APPROVED** [review]
lg, please address the nits before land

**Comment by danielvegamyhre** [comment]
> lg, please address the nits before land

thanks, done
